src/v2/auth.rs

```diff
@@ -1,23 +1,161 @@
 use crate::v2::*;
-use reqwest::{StatusCode, Url};
+use reqwest::{header::HeaderValue, RequestBuilder, StatusCode, Url};
+use std::iter::FromIterator;
+
+/// Represents all supported authentication schemes and is stored by `Client`.
+#[derive(Debug, Clone)]
+pub enum Auth {
+    Bearer(BearerAuth),
+    Basic(BasicAuth),
+}
+
+impl Auth {
+    /// Add authentication headers to a request builder.
+    pub(crate) fn add_auth_headers(&self, request_builder: RequestBuilder) -> RequestBuilder {
+        match self {
+            Auth::Bearer(bearer_auth) => request_builder.bearer_auth(bearer_auth.token.clone()),
+            Auth::Basic(basic_auth) => {
+                request_builder.basic_auth(basic_auth.user.clone(), basic_auth.password.clone())
+            }
+        }
+    }
+}
 
-#[derive(Debug, Default, Deserialize, Serialize)]
-pub struct TokenAuth {
+/// Used for Bearer HTTP Authentication.
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct BearerAuth {
     token: String,
     expires_in: Option<u32>,
     issued_at: Option<String>,
     refresh_token: Option<String>,
 }
 
-impl TokenAuth {
-    pub fn token(&self) -> &str {
-        self.token.as_str()
+/// Used for Basic HTTP Authentication.
+#[derive(Debug, Clone)]
+pub struct BasicAuth {
+    user: String,
+    password: Option<String>,
+}
+
+/// Structured representation for the content of the authentication response header.
+#[derive(Debug, PartialEq, Eq, Deserialize)]
+pub(crate) enum WwwAuthenticateHeaderContent {
+    Bearer(WwwAuthenticateHeaderContentBearer),
+    Basic(WwwAuthenticateHeaderContentBasic),
+}
+
+impl WwwAuthenticateHeaderContent {
+    /// Create a `WwwAuthenticateHeaderContent` by parsing a `HeaderValue` instance.
+    pub(crate) fn from_www_authentication_header(header_value: HeaderValue) -> Result<Self> {
+        let header = String::from_utf8(header_value.as_bytes().to_vec())?;
+
+        // This regex will result in multiple captures which will contain one key-value pair each.
+        // The first capture will be the only one with the "method" group set.
+        let re = regex::Regex::new(
+            r#"(?x)\s*
+            ((?P<method>[A-Z][a-z]+)\s*)?
+            (
+                \s*
+                    (?P<key>[a-z]+)
+                \s*
+                    =
+                \s*
+                    "(?P<value>[^"]+)"
+                \s*
+            )
+        "#,
+        )?;
+        let captures = re.captures_iter(&header).collect::<Vec<_>>();
+
+        let method = captures
+            .get(0)
+            .ok_or_else(|| {
+                Error::from(format!("regex '{}' didn't match '{}'", re.as_str(), header))
+            })?
+            .name("method")
+            .ok_or_else(|| Error::from(format!("method not found in {}", header)))?
+            .as_str();
+
+        let serialized_content = format!(
+            r#"{{ "{}": {{ {} }} }}"#,
+            method,
+            captures
```

Comment:
Nit: it would be a bit more readable having this iter-collection logic outside of the `format!()`.